### ISSUE TITLE
throttle doesn't work at all

### ISSUE THREAD

**Issue by r3wt** [SupremeTechnopriest/react-idle-timer#127]
First, love this package. There is just one small problem I am experiencing, and I am not sure why. 

code:
```jsx
const { getRemainingTime, getLastActiveTime } = useIdleTimer({
    timeout: 1000 * 20,// after 20s of activity, user is idle. next time user becomes active, will immediately call server
    onIdle: handleOnIdle,
    onActive: handleOnActive,
    onAction: handleOnAction,
    debounce: 0,
    throttle: 60000, // will only call server at most every 60 seconds for onAction
    eventsThrottle: 500, // prevent too much cpu usage for user actions
    capture: false,
    passive: true
  });
```
expected:
> onAction fires at most once every 60 seconds

actual:
> onAction fires every ~500ms.

**Comment by SupremeTechnopriest** [comment]
@r3wt Interesting!  I will take a look at this and add some tests.

**Comment by r3wt** [comment]
@SupremeTechnopriest thank you for fast response! If you are accepting PR's, i could possibly help if you are too busy. just let me know, i am happy to submit PR

**Comment by SupremeTechnopriest** [comment]
I accept PRs, but I should be able to get to this today.  No worries.  Just give me a few hours to get through some day job stuff and I'll take a look.

**Comment by SupremeTechnopriest** [comment]
@r3wt  I see whats happening here.  So `throttle` calls the onAction handler on the first action at most every `n` milliseconds. Events throttle is set to 500ms. so there will be a 500ms delay before it is called for the first time.  It sounds like you want the event to be called on the tail of the throttle? 

See here https://codesandbox.io/s/staging-frog-2sf04?file=/src/App.js

If you open the console and wiggle the mouse around continuously you will see that onAction is called once every 10000 ms on the head (front) of the action, so it is working as designed.  Even with constant input, an event will fire every `n` milliseconds. 

Debounce on the other hand will call onAction after `n` milliseconds of no activity.  Not sure this is what you want, but it does call `onAction` on the tail.  You could set debounce to something low like `500` and it would be called after 500ms of no activity. 

I could consider adding a configuration to call onAction on the tail of the throttle, but I dont think it should matter all that much as long as its only being called once every `n` milliseconds.  

Let me know what you think.

**Comment by SupremeTechnopriest** [comment]
@r3wt What are we thinking on this?  Will the current throttle implementation suffice for you?

**Comment by r3wt** [comment]
I'm not quite sure I understand what you are saying. it seems you may be getting different results than I am. What is happening for me is that onAction is calling every 500ms, and i don't understand why. i do not desire onAction call at tail of activity, but rather at leading edge as is. I just don't need it be called every 500ms, but instead at most every 1minute. i am wondering if this would be effected by rerender perhaps, causing the timeout to be initialized again on each render? this component is doing many different things related to the user session, so it is rerendering quite a bit, every second actually.

**Comment by SupremeTechnopriest** [comment]
If you can make a code sandbox that reproduces this we can take a look.  The `useIdleTimer` hook shouldn't be updating on re renders as it uses refs internally.  

Why don't you fork the example I made here  https://codesandbox.io/s/staging-frog-2sf04?file=/src/App.js and configure it the way you are using it. Just the bare minimum to reproduce the results you are getting should suffice.

**Comment by r3wt** [comment]
Sorry for not replying sooner. unfortunately, i can't use codesandbox(or any other such site) as it is blocked by corporate firewall. as for the problem, i realize now its probably that i am using this library in a way different than most use cases that this library was designed for, as my use case is simply that i don't care about user activity, i just want to renew the user session at most once per minute on leading edge on any user activity, but if user is not active at all, i do not want to renew session at all. so what i ended up doing is just wrapping my onAction/onIdle/onActive handlers using `throttle` from `throttle-debounce` package and everything is working fine now. apologies for long delay but i just didn't have an opportunity to work on this again until this week.

**Comment by roypearce** [comment]
I've had the same issue. Here is a sandbox that repros it https://codesandbox.io/s/boring-elbakyan-thhm6?file=/src/App.js Lose the code that is causing re-renders, and it works as expected. Since this is likely to be used on pages that have re-renders, this is a realistic issue. I imagine the cause is some sort of issue where a new instance is generated on each render? 

What I've found that solves for this use case is to use `eventsThrottle` alone.

**Comment by SupremeTechnopriest** [comment]
Thanks for the sandbox.  I will look into it this weekend!

**Comment by SupremeTechnopriest** [comment]
@roypearce I checked out the sandbox you linked and I dropped the timers down a bit to:

```javascript
const { getRemainingTime } = useIdleTimer({
    events: ["mousedown"],
    onAction: handleOnAction,
    onIdle: handleOnIdle,
    throttle: 1000,
    timeout: 2000
  });```

It seems to be working as designed.  With this config it will only allow `1` action event per every `1000ms`. Can you help me understand what you expect? Its worth noting that the events that reset the timer are separate from the `onAction` handler.  If you want to throttle events on the timer you would need to use `eventsThrottle`.

**Comment by roypearce** [comment]
Given the sandbox I shared, I expect one call to onAction every 5 seconds regardless of how many clicks happen on the page. In the real world I have this set to 5 minutes because onAction is making an API call to check if the session was modified, but I certainly don't want that to happen once a second.

Having the interval and the throttle set to the same amount will probably mask the issue. If you want to repro with your config then set the interval on line 13 to `100`. Then when you reload the example, with the console showing, immediately and rapidly click on the page. You'll notice you have 3+ onActions firing every second because there are multiple throttled functions being called which are generated from different renders. It would be easier to just go back to the original sandbox I shared which illustrates the issue quite rapidly without needing to modifying anything.

**Comment by SupremeTechnopriest** [comment]
Confirmed.  Will look into this tonight.

**Comment by SupremeTechnopriest** [comment]
Hey @roypearce @r3wt,

I found and fixed the issue.  Published to npm as `4.6.0-rc.2` under the `next` tag.  Install with `npm i react-idle-timer@next`.  Theres a few new features on that branch as well, but they are off by default.  Hoping to get this on the stable branch sometime in the next few weeks. 

Let me know if that fixes it for you. Looking good on my end.

**Comment by SupremeTechnopriest** [comment]
If you guys need this on the stable branch I can push a patch, just let me know if you cant wait a week and I'll get it done.

**Comment by r3wt** [comment]
@SupremeTechnopriest Thanks! I am ok to wait for patch later

**Comment by roypearce** [comment]
Thanks @SupremeTechnopriest for the fix. I think there's more than just us two that would benefit from it being patched. I appreciate the time & effort you put into maintaining this, and I'm glad you were able to find & fix the issue, so I'm okay waiting.

**Comment by SupremeTechnopriest** [comment]
Thanks guys!  Shouldn't be long.  Probably this Friday or this weekend.

**Comment by SupremeTechnopriest** [comment]
@roypearce @r3wt Hey guys, 4.6.0 is published to NPM.  The throttle bug is fixed and I also added some new features.  Check out the README for more info!

**Comment by roypearce** [comment]
@SupremeTechnopriest thank you sir!

**Comment by SupremeTechnopriest** [comment]
Hey @roypearce and @r3wt,

I made some discoveries while working on some new features in regards to throttle.  There is a trade off we need to decide on.  So basically if the component that mounts IdleTimer is functional, the callbacks `onActive`, `onIdle` and `onAction` are recreated every time the component updates.  This is why we were seeing that bug with the `throttle`.  The reason being is that we have the useIdleTimer set up so that you can redefine the event handlers.  We need to decide between being able to redefine handlers and wrapping them in `useCallback` or not being able to redefine them and being able to pass regular functions.

```javascript
const handlOnActive  = useCallback(() => console.log('active'), [])
idleTimer({ onActive: handleOnActive })

// vs

idleTimer({ onActive: () => console.log('active') })
```

I can make either implementation work, but if we want to keep the ability to dynamically reassign the event handlers, you would have to wrap them in `useCallback` at the application level.  What do you think?

**Comment by SupremeTechnopriest** [comment]
I am leaning towards keeping the ability to redefine and requiring the callbacks to be wrapped in `useCallback`.

**Comment by r3wt** [comment]
@SupremeTechnopriest I agree with you. Perhaps a note about the behavior in the readme (maybe a pitfalls section?) would be helpful.

**Comment by SupremeTechnopriest** [comment]
For sure, I will update the README and the examples.

**Comment by SupremeTechnopriest** [comment]
When 4.6.1 drops you will need to wrap your handlers in `useCallback`.  Just a heads up.

**Comment by SupremeTechnopriest** [comment]
Good news, I managed to get all the functionality we want without requiring `useCallback` so this wont be a breaking change, though I do still suggest it as it likely improves performance slightly.

**Comment by SupremeTechnopriest** [comment]
I published `4.6.1-rc.3` to npm under the `@next` tag if you want to test it out in your applications.  `npm i react-idle-timer@next`.

**Comment by SupremeTechnopriest** [comment]
`4.6.1`  is released to NPM.  You can switch back to the `latest` tag now.  You don't have to change your source code at all.  Managed to support anonymous, unwrapped callback functions.

**Comment by Slessi** [comment]
The fix from `4.6.1` seems to have stopped working since `5.0.0` - `useCallback` is required again

**Comment by SupremeTechnopriest** [comment]
> The fix from `4.6.1` seems to have stopped working since `5.0.0` - `useCallback` is required again

@Slessi can you provide a code sandbox that demonstrates the regression? All the tests that were written to cover this issue are still passing in v5.

**Comment by Slessi** [comment]
@SupremeTechnopriest https://codesandbox.io/s/relaxed-bash-oqfdbi?file=/src/App.js

Click on the screen, look at console, and you will see an `onAction` log every second, as my component re-renders every second, despite having `throttle` set to 1minute.

Then wrap the anonymous function on line 9 with `useCallback`, reload, and it will work again
```diff
-onAction: () => console.log(new Date()),
+onAction: useCallback(() => console.log(new Date()), []),
```

Same result for every `^5.0.0` version. Works fine in `~4.6.1`.

**Comment by SupremeTechnopriest** [comment]
@Slessi Just patched the regression in `5.4.2`.  Let me know if there its still an issue.

**Comment by Slessi** [comment]
@SupremeTechnopriest works now!
